src/google/protobuf/descriptor.cc

```diff
@@ -5487,6 +5497,84 @@ void DescriptorBuilder::BuildMessage(const DescriptorProto& proto,
   }
 }
 
+void DescriptorBuilder::CheckFieldJsonNameUniqueness(
+    const DescriptorProto& proto, const Descriptor* result) {
+  FileDescriptor::Syntax syntax = result->file()->syntax();
+  std::string message_name = result->full_name();
+  // two passes: one looking only at default JSON names, and one that considers
+  // custom JSON names
+  CheckFieldJsonNameUniqueness(message_name, proto, syntax, false);
+  CheckFieldJsonNameUniqueness(message_name, proto, syntax, true);
+}
+
+namespace {
+// Helpers for function below
+
+struct JsonNameDetails {
+  const FieldDescriptorProto* field;
+  std::string orig_name;
+  bool is_custom;
+};
+
+JsonNameDetails GetJsonNameDetails(const FieldDescriptorProto* field, bool use_custom) {
+  if (use_custom && field->has_json_name()) {
+    return {field, field->json_name(), true};
+  }
+  return {field, ToJsonName(field->name()), false};
+}
+
+
+} // namespace
+
+void DescriptorBuilder::CheckFieldJsonNameUniqueness(
+    const std::string& message_name, const DescriptorProto& message,
+    FileDescriptor::Syntax syntax, bool use_custom_names) {
+
+  absl::flat_hash_map<std::string, JsonNameDetails> name_to_field;
+  for (const FieldDescriptorProto& field : message.field()) {
+    JsonNameDetails details = GetJsonNameDetails(&field, use_custom_names);
+    std::string lowercase_name = absl::AsciiStrToLower(details.orig_name);
+    auto existing = name_to_field.find(lowercase_name);
+    if (existing == name_to_field.end()) {
+      name_to_field[lowercase_name] = details;
```

Comment:
this is a double map lookup.

```
auto it_inserted = named_to_field.try_emplace(lowercase_name, details);
if (it_inserted.second) {
  continue;
}
auto& match = it.first->second;
```

is the pattern to use to avoid it

Reply:
Ah, makes sense. Done.